core/src/main/java/com/netflix/msl/util/SimpleMslStore.java

```diff
@@ -108,19 +109,28 @@ public ICryptoContext getCryptoContext(final MasterToken masterToken) {
      */
     @Override
     public synchronized void removeCryptoContext(final MasterToken masterToken) {
-        if (cryptoContexts.remove(masterToken) != null) {
-            // Remove bound user ID tokens, service tokens, and the non-
-            // replayable ID if we no longer have a master token with the same
-            // serial number.
+        // We must perform the removal operations in reverse-dependency order.
+        // This ensures the store is in the correct state, allowing all logical
+        // and safety checks to pass.
+        //
+        // First any bound user ID tokens are removed (which first removes any
+        // service tokens bound to those user ID tokens), then bound service
+        // tokens, and finally the non-replayable ID and crypto context and
+        // master token pair.
+        if (cryptoContexts.containsKey(masterToken)) {
+            // Look for a second master token with the same serial number. If
```

Comment:
Isn't a bug that you could get into this situation in the first place, even during renewal? Why should it be allowed to add more than one CryptoContext for a given MT serial number?

Reply:
Because serial numbers don't change on master token renewal, only sequence numbers do. It is expected to have at least two pairs of a master token and crypto context during a renewal, and possibly more depending on the number of non-renewable parallel requests outstanding that haven't completed yet.